src/main/java/com/transferwise/common/gaffer/TransactionImpl.java

```diff
@@ -1,395 +1,439 @@
+  private AbandonedTransactionsTracker abandonedTransactionsTracker;
+
+  public TransactionImpl() {
+    ServiceRegistry serviceRegistry = ServiceRegistryHolder.getServiceRegistry();
+    String instanceId = serviceRegistry.getConfiguration().getInstanceId();
+    this.clock = serviceRegistry.getClock();
+    startTimeMillis = clock.currentTimeMillis();
+    globalTransactionId = new UidImpl(instanceId, startTimeMillis);
+    exceptionThrower = new ExceptionThrower(serviceRegistry.getConfiguration().isLogExceptions());
+
+    abandonedTransactionsTracker = new AbandonedTransactionsTracker(getTransactionManagerStatistics(), globalTransactionId, status);
+    cleaner.register(this, abandonedTransactionsTracker);
+  }
+
+  public void setSuspended(boolean suspended) {
+    this.suspended = suspended;
+  }
+
+  public void putResource(Object key, Object val) {
+    if (key == null) {
+      exceptionThrower.throwException(new IllegalArgumentException("Resource key can not be null."));
     }
+    resources.put(key, val);
+  }
 
-    public void begin(Integer timeoutSeconds, long beforeCommitValidationRequiredTimeMs) {
-        if (log.isDebugEnabled()) {
-            log.debug("Starting transaction '%s' with timeout of '%s' seconds.", getTransactionInfo(), timeoutSeconds == null ? "infinite" : timeoutSeconds);
-        }
-        if (timeoutSeconds != null) {
-            setTimeoutMillis(timeoutSeconds * 1000);
-        }
-        this.beforeCommitValidationRequiredTimeMs = beforeCommitValidationRequiredTimeMs;
-        status = Status.STATUS_ACTIVE;
+  public Object getResource(Object key) {
+    if (key == null) {
+      exceptionThrower.throwException(new IllegalArgumentException("Resource key can not be null."));
     }
+    return resources.get(key);
+  }
 
-    @Override
-    public void commit() throws RollbackException, IllegalStateException {
-        if (log.isDebugEnabled()) {
-            log.debug("Committing transaction '%s'.", getTransactionInfo());
-        }
-        if (status == Status.STATUS_NO_TRANSACTION) {
-            exceptionThrower.throwException(new IllegalStateException("Can not commit '" + getTransactionInfo() + "'. Transaction has not been started."));
-        }
-        if (isDoneOrFinishing()) {
-            exceptionThrower.throwException(new IllegalStateException("Can not commit '" + getTransactionInfo() + "' with status '" + status
-                + "''. Transaction is finishing or finished."));
-        }
+  public void begin(Integer timeoutSeconds, long beforeCommitValidationRequiredTimeMs) {
+    if (log.isDebugEnabled()) {
+      log.debug("Starting transaction '%s' with timeout of '%s' seconds.", getTransactionInfo(),
+          timeoutSeconds == null ? "infinite" : timeoutSeconds);
+    }
+    if (timeoutSeconds != null) {
+      setTimeoutMillis(timeoutSeconds * 1000);
+    }
+    this.beforeCommitValidationRequiredTimeMs = beforeCommitValidationRequiredTimeMs;
+    setStatus0(Status.STATUS_ACTIVE);
```

Comment:
This together with the other setStatus(Status) is very confusing, can we have another name for this?